File: Problemset/roman-to-integer/roman-to-integer.java

// @Title: 罗马数字转整数 (Roman to Integer)
// @Author: opfanswyx@outlook.com
// @Date: 2018-09-19 23:13:10
// @Runtime: 54 ms
// @Memory: N/A

class Solution {
    public int romanToInt(String s) {
        if(s==null||s.length()==0){
            return 0;
        }
        final Map<Character, Integer> m = new HashMap<>();
        m.put('I', 1);
        m.put('V', 5);
        m.put('X', 10);
        m.put('L', 50);
        m.put('C', 100);
        m.put('D', 500);
        m.put('M', 1000);
        
        int res = m.get(s.charAt(0));
        for(int i=1, len = s.length(); i<len;i++){
            int pre = m.get(s.charAt(i-1));
            int cur = m.get(s.charAt(i));
            
            res += cur - (cur<=pre?0:pre*2);
        }
        return res;
    }
}

File: Problemset/roman-to-integer/roman-to-integer.c

// @Title: 罗马数字转整数 (Roman to Integer)
// @Author: opfanswyx@outlook.com
// @Date: 2018-09-19 22:43:16
// @Runtime: 40 ms
// @Memory: N/A

int romanToInt(char* s) {
    int res = 0;
        while(*s!='\0')
        {
            switch(*s)
            {
                case 'M':
                    res += 1000;
                    s++;
                    break;
                case 'D':
                    res += 500;
                    s++;
                    break;
                case 'C':
                    (*(s+1)=='D'||*(s+1)=='M')?(res-=100):(res+=100);
                    s++;
                    break;
                case 'L':
                    res += 50;
                    s++;
                    break;
                case 'X':
                    (*(s+1)=='L'||*(s+1)=='C')?(res-=10):(res+=10);
                    s++;
                    break;
                case 'V':
                    res += 5;
                    s++;
                    break;
                case 'I':
                    (*(s+1)=='I'||*(s+1)==0)?(res++):(res--);
                    s++;
                    break;
            }
            
        }
    return res;
    
}
